src/theory/quantifiers/sygus/sygus_enumerator.cpp

```diff
@@ -0,0 +1,923 @@
+  children.push_back(Node::fromExpr(dt[cnum].getConstructor()));
+  // add the current of each child to children
+  for (unsigned i = 0, nargs = dt[cnum].getNumArgs(); i < nargs; i++)
+  {
+    Assert(d_children.find(i) != d_children.end());
+    children.push_back(d_children[i].getCurrent());
+  }
+  d_currTerm = NodeManager::currentNM()->mkNode(APPLY_CONSTRUCTOR, children);
+  return d_currTerm;
+}
+
+bool SygusEnumerator::TermEnumMaster::increment()
+{
+  // Am I already incrementing? If so, fail.
+  // This is to break infinite loops for slave enumerators who request an
+  // increment() from the master enumerator of their type that is also their
+  // parent. This ensures we do not loop on a grammar like:
+  //   A -> -( A ) | B+B
+  //   B -> x | y
+  // where we require the first term enumerated to be over B+B.
+  // Set that we are incrementing
+  if (d_isIncrementing)
+  {
+    return false;
+  }
+  d_isIncrementing = true;
+  bool ret = incrementInternal();
+  d_isIncrementing = false;
+  return ret;
+}
+
+bool SygusEnumerator::TermEnumMaster::incrementInternal()
+{
+  SygusEnumerator::TermCache& tc = d_se->d_tcache[d_tn];
+  if (tc.isComplete())
+  {
+    return false;
+  }
+  Trace("sygus-enum-debug2") << "master(" << d_tn
+                             << "): get last constructor class..." << std::endl;
+  // the maximum index of a constructor class to consider
+  unsigned ncc = tc.getLastConstructorClassIndexForWeight(d_currSize);
+  Trace("sygus-enum-debug2") << "Last constructor class " << d_currSize << ": "
+                             << ncc << std::endl;
+
+  // have we initialized the current constructor class?
+  while (d_ccCons.empty() && d_consClassNum < ncc)
+  {
+    Assert(d_ccTypes.empty());
+    Trace("sygus-enum-debug2") << "master(" << d_tn
+                               << "): try constructor class " << d_consClassNum
+                               << std::endl;
+    // get the list of constructors in the constructor class
+    tc.getConstructorClass(d_consClassNum, d_ccCons);
+    // if there are any...
+    if (!d_ccCons.empty())
+    {
+      // successfully initialized the constructor class
+      d_consNum = 0;
+      // we will populate the children
+      Assert(d_children.empty());
+      Assert(d_ccTypes.empty());
+      tc.getTypesForConstructorClass(d_consClassNum, d_ccTypes);
+      d_ccWeight = tc.getWeightForConstructorClass(d_consClassNum);
+      d_childrenValid = 0;
+      // initialize the children into their initial state
+      if (!initializeChildren())
+      {
+        // didn't work (due to size), we will try the next class
+        d_ccCons.clear();
+        d_ccTypes.clear();
+        Trace("sygus-enum-debug2") << "master(" << d_tn
+                                   << "): failed due to init size\n";
+      }
+    }
+    else
+    {
+      Trace("sygus-enum-debug2") << "master(" << d_tn
+                                 << "): failed due to no cons\n";
```

Comment:
what does it mean to get to this `else`?

Reply:
no constructors were in the constructor class.  This can only happen for the "nullary" constructor class for types that do not have nullary constructors.

Reply:
Maybe add this as a comment? I was wondering when it would be the case.

Reply:
Done.